web/data/accounts.go

```diff
@@ -0,0 +1,218 @@
+package data
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"net/url"
+
+	"github.com/cozy/cozy-stack/pkg/accounts"
+	"github.com/cozy/cozy-stack/pkg/consts"
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+	perms "github.com/cozy/cozy-stack/pkg/permissions"
+	"github.com/cozy/cozy-stack/web/jsonapi"
+	"github.com/cozy/cozy-stack/web/middlewares"
+	"github.com/cozy/cozy-stack/web/permissions"
+	"github.com/labstack/echo"
+)
+
+// TODO: make specific routes for managing accounts. The overriding of the
+// /data/io.cozy.accounts/* routes is here mainly for retro-compatible reasons,
+// but specific routes would improve the API.
+
+func accountDoctype(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		doctype := consts.Accounts
+		c.Set("doctype", doctype)
+		docidraw := c.Param("docid")
+		docid, err := url.QueryUnescape(docidraw)
+		if err != nil {
+			return jsonapi.NewError(http.StatusBadRequest, "Invalid docid '%s'", docid)
+		}
+		c.Set("docid", docid)
+		return next(c)
+	}
+}
+
+func getAccount(c echo.Context) error {
+	instance := middlewares.GetInstance(c)
+	doctype := consts.Accounts
+	docid := c.Get("docid").(string)
+	if docid == "" {
+		return dbStatus(c)
+	}
+
+	var out couchdb.JSONDoc
+	err := couchdb.GetDoc(instance, doctype, docid, &out)
+	if err != nil {
+		return fixErrorNoDatabaseIsWrongDoctype(err)
+	}
+	out.Type = doctype
+
+	if err = permissions.Allow(c, permissions.GET, &out); err != nil {
+		return err
+	}
+
+	if encryptAccount(out) {
+		if err = couchdb.UpdateDoc(instance, out); err != nil {
+			return err
+		}
+	}
+
+	perm, err := permissions.GetPermission(c)
+	if err != nil {
+		return err
+	}
+	if perm.Type == perms.TypeKonnector {
+		decryptAccount(out)
+	}
+
+	return c.JSON(http.StatusOK, out.ToMapWithType())
+}
+
+func updateAccount(c echo.Context) error {
+	instance := middlewares.GetInstance(c)
+
+	var doc couchdb.JSONDoc
+	if err := json.NewDecoder(c.Request().Body).Decode(&doc); err != nil {
+		return jsonapi.NewError(http.StatusBadRequest, err)
+	}
+
+	doc.Type = consts.Accounts
+
+	if (doc.ID() == "") != (doc.Rev() == "") {
+		return jsonapi.NewError(http.StatusBadRequest,
+			"You must either provide an _id and _rev in document (update) or neither (create with fixed id).")
+	}
+
+	if doc.ID() != "" && doc.ID() != c.Get("docid").(string) {
+		return jsonapi.NewError(http.StatusBadRequest, "document _id doesnt match url")
+	}
+
+	if doc.ID() == "" {
+		doc.SetID(c.Get("docid").(string))
+		return createNamedDoc(c, doc)
+	}
+
+	errWhole := permissions.AllowWholeType(c, permissions.PUT, doc.DocType())
+	if errWhole != nil {
+		// we cant apply to whole type, let's fetch old doc and see if it applies there
+		var old couchdb.JSONDoc
+		errFetch := couchdb.GetDoc(instance, doc.DocType(), doc.ID(), &old)
+		if errFetch != nil {
+			return errFetch
```

Comment:
Is it expected that we forbid the creation of a new account for permissions that are not on the whole type?

Reply:
Not really. I'll simplify this handler to only handle the whole type.

Reply:
Actually, it may be used by some konnectors who need to update their account during their execution. They have a specific permission on the account with its ID.